controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+			// label value length of 63 characters.
+			vm.Labels[vmopv1.VirtualMachineReplicaSetNameLabel] = util.MustFormatValue(ctx.ReplicaSet.Name)
+
+			return nil
+		})
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to apply %s label to VirtualMachine %q",
+				vmopv1.VirtualMachineReplicaSetNameLabel, vm.Name)
+		}
+
+		ctx.Logger.V(5).Info(
+			"Updated VirtualMachine with the VirtualMachineReplicaSet name label",
+			"result", res,
+			"vm", vm,
+			"key", vmopv1.VirtualMachineReplicaSetNameLabel,
+			"value", util.MustFormatValue(ctx.ReplicaSet.Name))
+
+		// TODO: Propagate the Deployment label from Deployment to replica set to VirtualMachine if
+		// it is set on the replica set.
+	}
+
+	// TODO: Handle in place propagation of fields to machines before syncing replicas.
+
+	syncErr := r.syncReplicas(ctx, ctx.ReplicaSet, filteredVMs)
+
+	// Update the status of the VirtualMachineReplicaSet even in case of error
+	// since syncing might have resulted in replicas being added or removed.
+	r.updateStatus(ctx, ctx.ReplicaSet, filteredVMs)
+
+	if syncErr != nil {
+		return ctrl.Result{}, errors.Wrapf(syncErr, "failed to sync VirtualMachineReplicaSet replicas")
+	}
+
+	var replicas int32
+	if ctx.ReplicaSet.Spec.Replicas != nil {
+		replicas = *ctx.ReplicaSet.Spec.Replicas
+	}
+
+	// Queue faster reconciles until all replicas are available.
+	if ctx.ReplicaSet.Status.ReadyReplicas != replicas {
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) (reterr error) {
-	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
-		defer func() {
-			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", reterr, false)
-		}()
+// MustEqualValue returns true if the replica set name equals either the label
+// value or its hashed value.
+func MustEqualValue(rsName, labelValue string) bool {
+	return labelValue == util.MustFormatValue(rsName)
+}
 
-		// Handle deletion of replicas
+// getNewVirtualMachine creates a new VirtualMachine object. The name of the newly created resource is going
+// to be created by the API server, we set the generateName field.
+func (r *Reconciler) getNewVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet) *vmopv1.VirtualMachine {
+	// gv := vmopv1.SchemeGroupVersion
+	vm := &vmopv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", rs.Name),
+			// Note: by setting the ownerRef on creation we signal to the
+			// VirtualMachine controller that this is not a stand-alone VM.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(rs, replicaSetKind)},
+			Namespace:       rs.Namespace,
+			Labels: map[string]string{
+				vmopv1.VirtualMachineReplicaSetNameLabel: util.MustFormatValue(rs.Name),
+			},
+			Annotations: rs.Spec.Template.Annotations,
+		},
+		// TypeMeta: metav1.TypeMeta{
```

Comment:
Is this supposed to be removed?

Reply:
Yep.  Intended to remove it after my testing.  Cleaned up.